### PR TITLE
feat: Updating the repository & OIDC module

### PR DESCRIPTION
**Pull request by gautambaghel**
- Added OIDC module to establish trust between Google Cloud and Terraform Cloud the dynamic JIT credentials eliminates the need for static secret keys
- Added more examples to the GKE module & the VM module
- Lot of quality of life fixes

### CODE REVIEW

File: examples/oidc-simple/main.tf [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+# The following variables must be set to allow runs
+# to authenticate to GCP.
+resource "tfe_variable" "enable_gcp_provider_auth" {
+  workspace_id = tfe_workspace.tfc_workspace.id
+
+  key      = "TFC_GCP_PROVIDER_AUTH"
+  value    = "true"
+  category = "env"
```
Comment (g-awmalik) [inline examples/oidc-simple/main.tf:65]:
```suggestion
  workspace_id = tfe_workspace.tfc_workspace.id
  key      = "TFC_GCP_PROVIDER_AUTH"
  value    = "true"
  category = "env"
```

File: examples/oidc-simple/main.tf [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+# TFC_GCP_WORKLOAD_PROVIDER_NAME variable contains the project number,
+# pool ID, and provider ID
+resource "tfe_variable" "tfc_gcp_workload_provider_name" {
+  workspace_id = tfe_workspace.tfc_workspace.id
+
+  key      = "TFC_GCP_WORKLOAD_PROVIDER_NAME"
+  value    = module.oidc.provider_name
+  category = "env"
+
+  description = "The workload provider name to authenticate against."
```
Comment (g-awmalik) [inline examples/oidc-simple/main.tf:79]:
```suggestion
  workspace_id = tfe_workspace.tfc_workspace.id
  key      = "TFC_GCP_WORKLOAD_PROVIDER_NAME"
  value    = module.oidc.provider_name
  category = "env"
  description = "The workload provider name to authenticate against."
```

File: examples/oidc-simple/main.tf [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+resource "tfe_variable" "tfc_gcp_service_account_email" {
+  workspace_id = tfe_workspace.tfc_workspace.id
+
+  key      = "TFC_GCP_RUN_SERVICE_ACCOUNT_EMAIL"
+  value    = google_service_account.sa.email
+  category = "env"
+
+  description = "The GCP service account email runs will use to authenticate."
```
Comment (g-awmalik) [inline examples/oidc-simple/main.tf:89]:
```suggestion
  workspace_id = tfe_workspace.tfc_workspace.id
  key      = "TFC_GCP_RUN_SERVICE_ACCOUNT_EMAIL"
  value    = google_service_account.sa.email
  category = "env"
  description = "The GCP service account TFC agents will use to authenticate."
```

File: examples/oidc-simple/variables.tf [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+variable "project_id" {
+  type        = string
+  description = "The project id to create WIF pool and example SA"
```
Comment (g-awmalik) [inline examples/oidc-simple/variables.tf:19]: Can we avoid using acronyms? Here and throughout especially WIF which is not well known.

File: examples/oidc-simple/versions.tf
Comment (g-awmalik) [inline examples/oidc-simple/versions.tf:1]: We can skip `versions.tf` for examples.

File: modules/tfc-agent-gke/main.tf [hunk 5]
```diff
@@ -119,3 +107,123 @@ resource "kubernetes_secret" "tfc_agent_secrets" {
...
+  }
+
+  spec {
+    scale_target_ref {
+      kind = "Deployment"
+      name = "${local.tfc_agent_name}-deployment"
+    }
+
+    min_replicas = 2
+    max_replicas = 10
```
Comment (g-awmalik) [inline modules/tfc-agent-gke/main.tf:214]: Can we make this configurable?

File: modules/tfc-agent-mig-vm/variables.tf [hunk 6]
```diff
@@ -16,30 +16,33 @@
...
   default     = "tfc-agent-network"
 }
 
-variable "create_network" {
-  type        = bool
-  description = "When set to true, VPC,router and NAT will be auto created"
-  default     = true
-}
-
 variable "subnetwork_project" {
```
Comment (g-awmalik) [inline modules/tfc-agent-mig-vm/variables.tf:40]:
```suggestion
variable "network_project" {
```

File: modules/tfc-oidc/versions.tf [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+terraform {
+  required_providers {
+    google = {
+      source  = "hashicorp/google"
+      version = ">= 3.53, < 5.0.0"
+    }
+  }
+  required_version = ">= 0.13"
```
Comment (g-awmalik) [inline modules/tfc-oidc/versions.tf:24]:
```suggestion
  required_version = ">= 0.13"
  
  provider_meta "google" {
    module_name = "blueprints/terraform/terraform-google-tf-cloud-agents:tfc-oidc/v0.0.1"
  }
```

Comment (g-awmalik) [reply modules/tfc-oidc/versions.tf:24]: Can we add these meta tags to all module version files? You can keep the version at 0.0.1 for now.

File: modules/tfc-oidc/variables.tf [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+variable "provider_description" {
+  type        = string
+  description = "Workload Identity Pool Provider description"
+  default     = "Workload Identity Pool Provider managed by Terraform"
+}
+
+
+# For more info please see
+# https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/iam_workload_identity_pool_provider
```
Comment (g-awmalik) [inline modules/tfc-oidc/variables.tf:69]: Can we add this to the description?

**Review by g-awmalik — CHANGES_REQUESTED** [review]
Thanks for the PR @gautambaghel. This is my first pass; will do another on Monday.

File: examples/oidc-simple/variables.tf (hunk 4 above)
Comment (gautambaghel) [reply examples/oidc-simple/variables.tf:19]: Done

File: modules/tfc-oidc/versions.tf (hunk 7 above)
Comment (gautambaghel) [reply modules/tfc-oidc/versions.tf:24]: Done!

File: modules/tfc-oidc/variables.tf (hunk 8 above)
Comment (gautambaghel) [reply modules/tfc-oidc/variables.tf:69]: Done!

File: modules/tfc-agent-gke/main.tf (hunk 5 above)
Comment (gautambaghel) [reply modules/tfc-agent-gke/main.tf:214]: Done!

**Pull request merged**